### PR TITLE
[WC-119] Implement Chef Service

### PR DESCRIPTION
**Pull request by Billy19191**
# Description

[//]: # 'Please include a summary of the changes and the related issue. (You can write below without deleting this comment)'

### What

[//]: # 'Briefly describe this feature / fix / refactor (You can write below without deleting this comment)'
I have introduce new chef service (microservice) and also inject to main gateway service such as route, business logic, and other stuffs.
*Please ignore the model at this stage; I will change it later once everything is stable.

### Why

[//]: # 'Describe the reason why we need this feature / fix / refactor or refer to Jira Ticket (You can write below without deleting this comment)'
To make service alive for chef routes and able to integration with frontend.

## Type of change

Please check the relevent option with `x`.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if needed)
- [x] My changes generate no new warnings
- [x] I have tested on my `Local Machine`

## Release Notes

- **New Features**
  - Introduced a new `ChefController` for managing chef profiles, including methods for signup, login, and profile management.
  - Added `ChefService` for handling chef-related operations via microservice communication.
  - Implemented structured error handling with `HttpExceptionFilter`.
  - Enhanced data modeling for chefs with the introduction of comprehensive interfaces and classes.
  - Updated the `README.md` to reflect the project focus on "WeCooked Backend."

- **Improvements**
  - Enhanced configuration management with the addition of `@nestjs/config`.
  - Improved logging and error handling in service methods.
  - Updated Webpack configuration for Hot Module Replacement (HMR) support.

- **Bug Fixes**
  - Removed deprecated methods and improved error handling in service methods.

- **Chores**
  - Updated dependencies in `package.json` for improved functionality.
  - Added `.gitignore` entries for sensitive files and directories.

### CODE REVIEW

File: gateway-service/prisma/migrations/20241108053414_wecooked/migration.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+  - You are about to drop the column `chefUsername` on the `Chef` table. All the data in the column will be lost.
+
+*/
+-- DropIndex
+DROP INDEX "Chef_chefUsername_key";
+
+-- AlterTable
+ALTER TABLE "Chef" DROP COLUMN "chefUsername",
+ADD COLUMN     "chefSex" TEXT;
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/prisma/migrations/20241108053414_wecooked/migration.sql:12]: Ignore this, base on what we love gender equality (LGBTQ support), so genders are string!

File: gateway-service/src/chef/chef.controller.spec.ts [hunk 2]
```diff
@@ -1,17 +1,17 @@
 import { Test, TestingModule } from '@nestjs/testing';
-import { AppController } from './app.controller';
-import { AppService } from './app.service';
+import { ChefController } from './chef.controller';
+import { ChefService } from './chef.service';
 
 describe('AppController', () => {
-  let appController: AppController;
+  let appController: ChefController;
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/chef/chef.controller.spec.ts:6]: Ignore all unit test files.

File: chef-service/src/http-exception.filter.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    const response = ctx.getResponse<Response>();
+    const request = ctx.getRequest<Request>();
+    const status = exception.getStatus();
+    this.logger.error(`Error: ${exception.message}`, exception.stack);
+    response.status(status).json({
+      statusCode: status,
+      timestamp: new Date().toISOString(),
+      path: request.url,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/http-exception.filter.ts:24]: No need, just leave it as it.

File: chef-service/src/main.ts [hunk 4]
```diff
@@ -17,6 +17,8 @@ async function bootstrap() {
     },
   );
   app.listen();
-  logger.log('Microservice is listening');
+  app.useGlobalFilters(new HttpExceptionFilter());
+  // app.useGlobalFilters(new HttpExceptionFilter());
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/main.ts]: Fixed this.

File: gateway-service/src/http-exception.filter.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  ArgumentsHost,
+  HttpException,
+  Logger,
+  HttpStatus,
+} from '@nestjs/common';
+import { Request, Response } from 'express';
+
+@Catch(HttpException)
+export class HttpExceptionFilter implements ExceptionFilter {
+  logger = new Logger('Gateway Service');
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/http-exception.filter.ts:12]: Ignore this, it's more readability

File: gateway-service/src/main.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+import { NestExpressApplication } from '@nestjs/platform-express';
+import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
+import { ValidationPipe } from '@nestjs/common';
+import { HttpExceptionFilter } from './http-exception.filter';
+
+declare const module: any;
+async function bootstrap() {
+  const app = await NestFactory.create<NestExpressApplication>(AppModule);
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/main.ts:10]: Ignore this.

File: gateway-service/src/app.module.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import {
+  HttpException,
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/app.module.ts]: Fixed.

File: gateway-service/src/app.module.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  imports: [
+    ConfigModule.forRoot({ isGlobal: true }),
+    ClientsModule.register([
+      {
+        name: 'CHEF_SERVICE',
+        transport: Transport.NATS,
+        options: {
+          servers: ['nats://localhost:4222'],
+        },
+      },
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/app.module.ts:20]: Leave it as it for now

File: chef-service/src/app.controller.ts [hunk 9]
```diff
@@ -1,12 +1,71 @@
...
+  @EventPattern('chef')
+  @EventPattern('chef/signup')
+  async postSignUpChef(@Body() body: any[]) {
+    try {
+      const data = await this.appService.postSignUpChef(body);
+      return data;
+    } catch (error) {
+      return error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.controller.ts:17]: As I said, Ignore all model for now

File: chef-service/package.json [hunk 10]
```diff
@@ -21,9 +21,12 @@
   },
   "dependencies": {
     "@nestjs/common": "^10.0.0",
+    "@nestjs/config": "^3.3.0",
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/package.json:24]: You're wrong, this is current version.

File: README.md [hunk 11]
```diff
@@ -5,24 +5,7 @@
...
-    <a href="https://opencollective.com/nest#sponsor"  target="_blank"><img src="https://img.shields.io/badge/Support%20us-Open%20Collective-41B883.svg" alt="Support us"></a>
-  <a href="https://twitter.com/nestframework" target="_blank"><img src="https://img.shields.io/twitter/follow/nestframework.svg?style=social&label=Follow" alt="Follow us on Twitter"></a>
-</p>
-  <!--[![Backers on Open Collective](https://opencollective.com/nest/backers/badge.svg)](https://opencollective.com/nest#backer)
-  [![Sponsors on Open Collective](https://opencollective.com/nest/sponsors/badge.svg)](https://opencollective.com/nest#sponsor)-->
-
-## Description
+## WeCooked Backend
 
 [Nest](https://github.com/nestjs/nest) framework TypeScript starter repository.
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply README.md:10]: Ignore this.

File: gateway-service/src/chef/chef.service.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  async updateProfileChef(body: ChefModel, id: number): Promise<any> {
+    try {
+      const result = await lastValueFrom(
+        this.chefClient.send('chef/updateProfile', { body, id }),
+      );
+      return result;
+    } catch (error) {
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/chef/chef.service.ts:66]: We will add model and interface later.

File: gateway-service/src/chef/chef.service.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  getChef(): Observable<any> {
+    return new Observable((observer) => {
+      lastValueFrom(this.chefClient.send('chef', {}))
+        .then((result) => {
+          observer.next(result);
+          observer.complete();
+        })
+        .catch((error) => observer.error(error));
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/chef/chef.service.ts]: Fixed, remove this as it is unused

File: chef-service/src/app.service.ts [hunk 14]
```diff
@@ -1,8 +1,240 @@
...
+      if (e instanceof Prisma.PrismaClientKnownRequestError) {
+        // Prisma error code for unique constraint violation
+        if (e.code === 'P2002') {
+          throw {
+            status: HttpStatus.CONFLICT,
+            message: 'Email already registered',
+          };
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.service.ts]: Handled this.

File: chef-service/src/app.service.ts [hunk 15]
```diff
@@ -1,8 +1,240 @@
-import { Injectable } from '@nestjs/common';
+import { Injectable, OnModuleInit, Logger, HttpStatus } from '@nestjs/common';
+import {
+  PrismaClient,
+  Prisma,
+} from '../../gateway-service/node_modules/.prisma/client';
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.service.ts:5]: No, we share from gateway-service as Prisma already generate at that folder.

File: gateway-service/src/middleware/logger.middleware.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+import { Request, Response, NextFunction } from 'express';
+
+@Injectable()
+export class LoggerMiddleware implements NestMiddleware {
+  logger = new Logger('Chef Service');
+  use(req: Request, res: Response, next: NextFunction) {
+    this.logger.log('Request...');
+    next();
+  }
+}
```
[the remark this answers is not in the document]
Comment (AungKhantThaw47) [reply gateway-service/src/middleware/logger.middleware.ts]: can upgrade to see req body and url

File: gateway/package.json
Comment (AungKhantThaw47) [inline gateway/package.json:1]: can I know why we need update for webpack?

File: gateway/webpack-hmr.config.js
Comment (AungKhantThaw47) [inline gateway/webpack-hmr.config.js:1]: can you elaborate more about this?

File: gateway/package.json
Comment (Billy19191) [reply gateway/package.json:1]: Because I config gateway service able to hot-reload which need webpack dependency.

File: gateway/webpack-hmr.config.js
Comment (Billy19191) [reply gateway/webpack-hmr.config.js:1]: As I mentioned above, please see this for reference [https://docs.nestjs.com/recipes/hot-reload](https://docs.nestjs.com/recipes/hot-reload)

File: gateway-service/src/chef/chef.service.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+  async uploadCourseVideo(id: number, payload: any): Promise<any> {
+    try {
+      const result = await lastValueFrom(
+        this.chefClient.send('chef/uploadCourseVideo', { id, payload }),
+      );
+      return result;
+    } catch (error) {
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/chef/chef.service.ts:77]: Ignore about interface and type at this moment.

[the remark this answers is not in the document]
Comment (Billy19191) [reply gateway-service/src/chef/chef.service.ts:77]: Ignore about checking / validate request for now, focus on core functional. Will fix this later

File: chef-service/src/app.service.ts [hunk 18]
```diff
@@ -1,8 +1,246 @@
...
+    const {
+      chefName,
+      chefSurname,
+      chefEmail,
+      chefPassword,
+      chefBio,
+      chefExperience,
+      chefSpecialty,
+      chefPhone,
+    }: ChefModel = body[0];
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.service.ts:31]: Not at this stage

File: chef-service/src/app.service.ts [hunk 19]
```diff
@@ -1,8 +1,246 @@
...
+        if (e.code === 'P2025') {
+          throw {
+            status: HttpStatus.NOT_FOUND,
+            message: 'Incorrect email or password',
+          };
+        }
+      }
+      throw e;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.service.ts:95]: We discuss about this later.

File: chef-service/src/app.service.ts [hunk 20]
```diff
@@ -1,8 +1,246 @@
...
+          status: HttpStatus.NOT_FOUND,
+          message: 'Chef not found',
+          data: [],
+        };
+      }
+      return {
+        status: HttpStatus.OK,
+        message: 'Chef found',
+        data: chef,
+      };
```
[the remark this answers is not in the document]
Comment (Billy19191) [reply chef-service/src/app.service.ts]: Fixed.

**Review by AungKhantThaw47 — APPROVED** [review]
Good to go

**Pull request merged**